src/Microsoft.Azure.WebJobs.Host/Bindings/BindingProviders/GenericDynamicAsyncCollectorBindingProvider.cs

```diff
@@ -0,0 +1,131 @@
+            this._converterManager = converterManager;
+            this._asyncCollectorType = asyncCollectorType;
+            this._constructorParameterBuilder = constructorParameterBuilder;
+        }
+
+        // Called once per method definition. Very static. 
+        public Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            ParameterInfo parameter = context.Parameter;
+            TAttribute attribute = parameter.GetCustomAttribute<TAttribute>(inherit: false);
+
+            if (attribute == null)
+            {
+                return Task.FromResult<IBinding>(null);
+            }
+
+            // Now we can instantiate against the user's type.
+            // throws if can't infer the type. 
+            Type typeMessage = TypeUtility.GetMessageTypeFromAsyncCollector(parameter.ParameterType);
+
+            var wrapper = WapperBase.New(
+                typeMessage, _asyncCollectorType, _constructorParameterBuilder, _nameResolver, _converterManager, parameter);
+
+
+            IBinding binding = wrapper.CreateBinding();
+            return Task.FromResult(binding);
+        }
+
+        // Wrappers to help with binding to a dynamically typed IAsyncCollector<T>. 
+        // TMessage is not known until runtime, so we need to dynamically create it. 
+        // These inherit the generic args of the outer class. 
+        abstract class WapperBase
+        {
+            protected Func<TAttribute, TConstructorArg> _constructorParameterBuilder;
+            protected INameResolver _nameResolver;
+            protected IConverterManager _converterManager;
+            protected ParameterInfo _parameter;
+            protected Type _asyncCollectorType;
+
+            public abstract IBinding CreateBinding();
+
+            internal static WapperBase New(
+                Type typeMessage,
+                Type asyncCollectorType,
+                Func<TAttribute, TConstructorArg> constructorParameterBuilder,
+                INameResolver nameResolver,
+                IConverterManager converterManager,
+                ParameterInfo parameter
+                )
+            {
+                // These inherit the generic args of the outer class. 
+                var t = typeof(Wapper<>).MakeGenericType(typeof(TAttribute), typeof(TConstructorArg), typeMessage);
+                var obj = Activator.CreateInstance(t);
+                var obj2 = (WapperBase)obj;
+
+                obj2._constructorParameterBuilder = constructorParameterBuilder;
+                obj2._nameResolver = nameResolver;
+                obj2._converterManager = converterManager;
+                obj2._parameter = parameter;
+                obj2._asyncCollectorType = asyncCollectorType;
+
+                return obj2;
+            }
+        }
+
+        class Wapper<TMessage> : WapperBase
```

Comment:
To clarify: Are you suggesting I rename WrapperBase to Wrapper? 

---
In reply to: [61139587](https://github.com/Azure/azure-webjobs-sdk/pull/694#discussion_r61139587) [](ancestors = 61139587)